### PR TITLE
MRG: Extract measurement date and age for NIRX files

### PR DESCRIPTION
**Pull request by rob-luke**
#### What does this implement/fix?
Extract measurement date and birthday for NIRX files.

#### Additional information
The vendor software only allows saving of age as an integer. So I have taken the measurement date, subtracted the age integer from the years, and stored that as the birthday. Open to alternative suggestions.

### CODE REVIEW

**Comment by rob-luke** [comment]
@larsoner this is failing due to processing of French date strings. Is there any inbuilt multilingual date string parsers in MNE already or do I need to write one?

**Comment by rob-luke** [comment]
HHmmmm my changes also seem to break other things, Ill fix that. But still curious about the best way to process French dates.

**Comment by agramfort** [comment]
what do you mean French dates? like month with letter with accents?

**Comment by larsoner** [comment]
Like Travis says:
```
ValueError: time data '"jeu. 13 févr. 2020""09:08:47.511"' does not match format '"%a, %b %d, %Y""%H:%M:%S.%f"'
```

**Comment by agramfort** [comment]
is it fixable by setting the LOCALE properly?

>

**Comment by larsoner** [comment]
The docs suggest that `locale` is used, so indeed if the `locale` can be extracted from the file header, then we can make a `@contextlib.contextmanager` to `orig = locale.set_locale(locale.LC_something, whatever)` and then `set_locale(locale.LC_something, orig)` afterward. Even if it isn't embedded in the header (which would be sad), we could start making a list of ones to iterate over and try, and always add others later if need be. (Or maybe it won't be so bad to iterate over all of them, if we can get a list?)

**Comment by rob-luke** [comment]
Thanks for the helpful suggestions. The locale is not embedded in the header. I had started writing a list of locales to iterate over, that's when I stopped and decided to ask here before progressing too far. So it sounds like there is no solution in MNE for this already, so I will continue to code along this path.

**Comment by larsoner** [comment]
Indeed AFAIK we have never had to deal with locale when parsing datetime. We can try adding some other `LC_` env var to some Travis or Azure run to see if we can get some existing `datetime` stuff to break. Let's do that after you fix the reading of this particular file, though 
-- it will probably tell us which env var(s) are relevant.

**Comment by agramfort** [comment]
here is an example that works:

import locale
from datetime import datetime
locale.setlocale(locale.LC_TIME, "fr_FR")
s = "jeu. 13 fév. 2020""09:08:47.511"
f = "%a. %d %b. %Y""%H:%M:%S.%f"
datetime.strptime(s, f)

however I had to use fév. and févr. which seems to be the standard.

**Comment by rob-luke** [comment]
> however I had to use fév. and févr. which seems to be the standard.

Thanks for the code snippet. I came up with something very similar and ran in to the same problem. It's the `févr` that is the problem, `datetime.strptime` seems to expect `fev`. I considered some sort of replace approach for different languages, but this seems unwieldily (how many languages to I then support?). I also considered other packages that are made for dealing with dates, but don't want to add more requirements.

**Comment by rob-luke** [comment]
Next Im going to check where this `févr` actually comes from. This may be a problem of our own creating, I have never seen a file from France and don't know if the device will actually save as `fev` or `févr`, this seems to have been[ generated somewhere in `_test_raw_reader`.](https://github.com/mne-tools/mne-python/blob/e69e581d58bdeada61c8b863885387aa28ede1da/mne/io/tests/test_raw.py#L43)

**Comment by agramfort** [comment]
I see with company that produces this file how the come up with févr while
it appears the standard across languages is fév

I would use  try except here to avoid failing to load such files but I
agree the mne code cannot fix all IO corner cases.

can you see with them how it happens? and maybe how they deal with this
internally?

>

**Comment by larsoner** [comment]
@rob-luke will you have time to revisit this in the next couple of weeks? I'll optimistically mark for 0.21

**Comment by rob-luke** [comment]
Yes I would like to cross this off. The milestone has sept 15 marked, that should be achievable

**Comment by rob-luke** [comment]
@larsoner I am not sure what to do here with the non standard French encoding of dates.

It seems to me that `fév` is what datetime expects, yet we have 'févr' in the tests because that was provided in #7313 (I assume, I don't actually have the file).

So if NIRX is providing non standard french date formats, and if I know all the non standard naming, I could create a translation function. But I don't know all the non standard naming and don't have access to a machine at the moment to determine this. Plus this would become quite unwieldy, what other non standard language formats do we then support.

Would it be acceptable to try and parse the dates in the current `locale` and throw a warning if it doesn't work. For dates that we can't extract it will set them to 0 or something else? This will get English working and downstream usage of the date, and when someone with access to a French system gets some data they can just tweak the French parsing. Thoughts?

**Comment by agramfort** [comment]
yes I fear we need something custom. But I would reach out to
engineers at the company
that produces these files to report the issue.

**Comment by rob-luke** [comment]
> I would reach out to engineers at the company that produces these files to report the issue.

Good idea. But I would like to verify the issue myself first if I am to contact them. Ideally someone could send me some files recorded in France. Failing that, I could set the local of our lab pc to France and record my own file, but Im currently locked out of the lab (at least for the next 3 weeks). So this won't be done by 0.21 release date. I suggest I create an English date PR for 0.21 and raise an issue for the non English dates to be solved when someone sends me some files or I can get in the lab. Acceptable? Or do you prefer I wait and do everything at once?

**Comment by agramfort** [comment]
acceptable

**Comment by rob-luke** [comment]
@agramfort @larsoner can you please review.

File: mne/io/nirx/nirx.py [hunk 1]
```diff
@@ -222,7 +253,8 @@ def __init__(self, fname, preload=False, verbose=None):
         def prepend(list, str):
             str += '{0}'
             list = [str.format(i) for i in list]
-            return(list)
+            return (list)
```
Comment (agramfort) [inline mne/io/nirx/nirx.py:256]:
```suggestion
            return list
```

besides you should not call a variable list

File: mne/io/nirx/tests/test_nirx.py [hunk 2]
```diff
@@ -79,6 +79,12 @@ def test_nirx_15_2_short():
     # Test data import
     assert raw._data.shape == (26, 145)
     assert raw.info['sfreq'] == 12.5
+    assert raw.info['meas_date'].month == 8
+    assert raw.info['meas_date'].day == 23
+    assert raw.info['meas_date'].year == 2019
+    assert raw.info['meas_date'].hour == 7
+    assert raw.info['meas_date'].minute == 37
+    assert raw.info['meas_date'].second == 4
```
Comment (agramfort) [inline mne/io/nirx/tests/test_nirx.py:87]: I would do it with == after creating a datetime instance with proper timezone. It's less much lines.

File: mne/io/nirx/tests/test_nirx.py [hunk 3]
```diff
@@ -268,16 +278,24 @@ def test_nirx_15_2():
     # Test data import
     assert raw._data.shape == (64, 67)
     assert raw.info['sfreq'] == 3.90625
+    assert raw.info['meas_date'].month == 10
+    assert raw.info['meas_date'].day == 2
+    assert raw.info['meas_date'].year == 2019
+    assert raw.info['meas_date'].hour == 9
+    assert raw.info['meas_date'].minute == 8
+    assert raw.info['meas_date'].second == 47
```
Comment (agramfort) [inline mne/io/nirx/tests/test_nirx.py:286]: same here one assert should do it

File: mne/io/nirx/tests/test_nirx.py [hunk 4]
```diff
@@ -312,6 +330,12 @@ def test_nirx_15_0():
     # Test data import
     assert raw._data.shape == (20, 92)
     assert raw.info['sfreq'] == 6.25
+    assert raw.info['meas_date'].month == 10
+    assert raw.info['meas_date'].day == 27
+    assert raw.info['meas_date'].year == 2019
+    assert raw.info['meas_date'].hour == 13
+    assert raw.info['meas_date'].minute == 53
+    assert raw.info['meas_date'].second == 34
```
Comment (agramfort) [inline mne/io/nirx/tests/test_nirx.py:338]: idem

File: mne/io/nirx/nirx.py [hunk 5]
```diff
@@ -118,22 +118,51 @@ def __init__(self, fname, preload=False, verbose=None):
...
+            meas_date = dt.datetime.strptime(datetime_str, dt_code)
+            meas_date = meas_date.replace(tzinfo=dt.timezone.utc)
+        except ValueError:
+            pass
+        try:
+            dt_code = '"%a, %d %b %Y""%H:%M:%S.%f"'
+            meas_date = dt.datetime.strptime(datetime_str, dt_code)
+            meas_date = meas_date.replace(tzinfo=dt.timezone.utc)
+        except ValueError:
+            pass
```
Comment (agramfort) [inline mne/io/nirx/nirx.py:136]:
```suggestion
        try:
            for dt_code in ['"%a, %b %d, %Y""%H:%M:%S.%f"', '"%a, %d %b %Y""%H:%M:%S.%f"']:
                meas_date = dt.datetime.strptime(datetime_str, dt_code)
                meas_date = meas_date.replace(tzinfo=dt.timezone.utc)
                break
        except ValueError:
            pass
```

**Comment by rob-luke** [comment]
Thanks for the review. @agramfort and @larsoner could you please review again.
CI fail seems to be a time out error

**Review by agramfort — APPROVED** [review]

**Comment by agramfort** [comment]
thx @rob-luke

**Pull request merged**